tech: integration of collector (#598)
sumup-oss/circuit-ui: docs/advanced/tracking.story.mdx

import { Meta, Intro, Story } from '../../.storybook/components';

<Meta title="Advanced/Event Tracking" />

# Event Tracking

<Intro>

>   Tracking user interactions enables you to learn about usage patterns and

Reviewer: ```suggestion
  Tracking user interactions enables you to learn about usage patterns.
```

>   improve the application. Circuit UI makes this straight-forward by integrating

Reviewer: ```suggestion
Circuit UI provides out of the box interaction tracking by leveraging <a href="https://github.com/sumup-oss/collector">Collector</a>. 

In this guide we will cover all the basics to start measuring using Circuit UI components. We highly recommending checking <a href="https://github.com/sumup-oss/collector">Collector</a> out for an in-depth overview.
```

  <a href="https://github.com/sumup/collector">Collector</a>. Here are the steps
  to enable and use it.
</Intro>

## Configuring tracking dispatch

1. Wrap your application in root level with `TrackingRoot` and provide a dispatch function for your tracking service.
2. Add `TrackingView` to a page or other meaningful view (e.g. a modal) in your application. (For more explanation about event schema, please refer to [Schema](https://github.com/sumup-oss/collector#schema) in Collector.)
3. To ensure high data quality, the `label` tracking property is mandatory in Collector. Circuit UI components invoke the dispatch function only when a value for the `label` property is passed as part of the `tracking` prop.
4. When a user interacts with a component, Circuit UI invokes the dispatch function with the standard event schema.

```jsx
import { TrackingRoot, TrackingView } from '@sumup/collector';

export default function App() {

>   const sendTracking = payload => sendGA(payload);

Reviewer: [in collector](https://github.com/sumup/collector#trackingroot) we recommend memoizing the dispatchHandler to avoid unnecessary renders. I think we should reflect it here

import { Meta, Intro, Story } from '../../.storybook/components';

<Meta title="Advanced/Build-in Tracking" />

# Build-in Tracking

<Intro>
  Working along with{' '}

>   <a href="https://github.com/sumup-oss/collector">Collector</a>, button-like
>   components and meanful UI operation in Circuit UI will come along with

Reviewer 2: ```suggestion
  <a href="https://github.com/sumup-oss/collector">Collector</a>, interactive
  components in Circuit UI will come along with
```

Author: I'm thinking about components like "tabs" and "pagination". If we want to track the interaction every time a user clicks on it, what `label` should we set as default? For example, we have 3 tabs rendered, should we set different `label` for each of these 3 different tabs? 
We could also not set any default tracking and leave this to the consumer. They will add the tracking by themselves. In this case, we only add default tracking to "button-like" component and not every interactive component. That's why I put the word here.

Reviewer 2: I was under the assumption that we don't provide a default `label`, only a default `component` value. The consumer needs to pass at least a `label` to enable tracking. 

This way, we just need to ensure that unique labels can be set on the Tabs and Pagination subcomponents.

Author: Ahhh true! I messed up the concept of component and label. Got it. thank you!

  build-in tracking method. Components like Button, Tag and Hambuger will invoke
  a pre-set dispatch function to trigger tracking event. For more detail

>   concept, please refer to our collector project.

Reviewer 2: Documentation should be written in present tense, not future tense, since it describes the state-is not as it will be.

</Intro>

## Configuring tracking dispatch

1. Wrap your application in root level with `TrackingRoot` and provide a dispatch function for your tracking event.
2. Add `TrackingView` to the viewport that make sense to your application. (For more concept explanation about `TrackingRoot and TrackingView`, please refer to [Collector](https://github.com/sumup-oss/collector).)
3. For better data quality, Collector set `label` as a mandatory field. Component will invoke dispatch function only when a label is provided.

> 4. When user clicks a button, a pre-set dispatch function will be invoked with parametres.

Reviewer 2: Point 3 mentions that the `label` is required, however I'm missing a note on how this can be done (through the `tracking` prop).

```jsx
import { TrackingRoot, TrackingView } from '@sumup/collector';

export default function App() {
  const sendTracking = payload => sendGA(payload);

  return (
    <TrackingRoot name="tracking-root" onDispatch={sendTracking)}>
      <TrackingView name="tracking-view">
        <Button
          tracking={{
            label: 'trackingId',
            component: 'button',
            customParameters: { custom1: 'data' }
          }}
        >
          {'Click'}
        </Button>
      </TrackingView>
    </TrackingRoot>
  );
}
```

## Components with build-in tracking

### Button Click

<Story id="components-button--tracking" />

### Clickable/Removable Tag

<Story id="components-tag--clickable" />

### Hambuger open/close

<Story id="components-hamburger--tracking" />

### Calendar Tag

<Story id="components-calendar-calendartag--base" />

### Calendar Tag Two Step

> <Story id="components-calendar-calendartagtwostep--base" />

Reviewer 2: I don't think it's necessary to maintain a list of the components that support tracking. The general rule is: if it's interactive, it supports tracking. Plus, users can check the prop types of a component to see whether it accepts the tracking prop or submit a PR if it doesn't. 

This list would be prone to become outdated.

import { Meta, Intro, Story } from '../../.storybook/components';

<Meta title="Advanced/Event Tracking" />

# Event Tracking

<Intro>
Tracking user interactions enables you to learn about usage patterns.
Circuit UI provides out of the box interaction tracking by leveraging <a href="https://github.com/sumup-oss/collector">Collector</a>. 

In this guide we will cover all the basics to start measuring using Circuit UI components. We highly recommending checking <a href="https://github.com/sumup-oss/collector">Collector</a> out for an in-depth overview. Here are the steps to enable and use it.
</Intro>

## Configuring Tracking Dispatch

1. Wrap your application in root level with `TrackingRoot` and provide a dispatch function for your tracking service.

> 2. Add `TrackingView` to a page or other meaningful view (e.g. a modal) in your application. (For more explanation about event schema, please refer to [Schema](https://github.com/sumup-oss/collector#schema) in Collector.)

Reviewer 2: ```suggestion
2. Add `TrackingView` to a [page view](https://github.com/sumup-oss/collector#page-view) in your application. Read more about the [Event Schema](https://github.com/sumup-oss/collector#event-schema) in the Collector documentation.
```
